src/train.py

```diff
@@ -100,113 +100,6 @@ def train_model(settings, hyp_params, train_loader, valid_loader, test_loader):
         else:
             model = model.cuda()
 
-    def train(model, optimizer, criterion, local_rank, reg_en=False):
-
-        epoch_loss = 0
-        model.train()
-        num_batches = hyp_params.n_train // hyp_params.batch_size
-        proc_loss, proc_size = 0, 0
-        start_time = time.time()
-
-        for i_batch, (batch_X, batch_Y, batch_META) in enumerate(train_loader):
-            sample_ind, text, audio, vision = batch_X
-            eval_attr = batch_Y.squeeze(-1)  # if num of labels is 1
-
-            model.zero_grad()
-            optimizer.zero_grad()
-
-            if hyp_params.use_cuda:
-                # with torch.cuda.device(0):
-                if hyp_params.ddp:
-                    text, audio, vision, eval_attr = text.cuda(local_rank), audio.cuda(local_rank), vision.cuda(
-                        local_rank), eval_attr.cuda(local_rank)
-                else:
-                    text, audio, vision, eval_attr = text.cuda(), audio.cuda(), vision.cuda(), eval_attr.cuda()
-
-            batch_size = text.size(0)
-
-            combined_loss = 0
-            preds, hiddens, h_list = model(text, audio, vision)
-
-            raw_loss = criterion(preds, eval_attr)
-
-            if reg_en:
-                # Sinkhorn Distances
-                # reg_term = torch.mul(h_list[-1], 0.1)
-                reg_term = h_list[-1]
-                combined_loss += hyp_params.reg_lambda * torch.mean(reg_term)
-            else:
-                combined_loss = raw_loss
-
-            combined_loss.backward()
-
-            torch.nn.utils.clip_grad_norm_(model.parameters(), hyp_params.clip)
-            optimizer.step()
-
-            proc_loss += raw_loss.item() * batch_size
-            proc_size += batch_size
-            epoch_loss += combined_loss.item() * batch_size
-            if i_batch % hyp_params.log_interval == 0 and i_batch > 0:
-                avg_loss = proc_loss / proc_size
-                elapsed_time = time.time() - start_time
-                print('Epoch {:2d} | Batch {:3d}/{:3d} | Time/Batch(ms) {:5.2f} | Train Loss {:5.4f}'.
-                      format(epoch, i_batch, num_batches, elapsed_time * 1000 / hyp_params.log_interval, avg_loss))
-                proc_loss, proc_size = 0, 0
-                start_time = time.time()
-
-            del text, audio, vision, eval_attr, preds, hiddens, h_list
-            gc.collect()
-            torch.cuda.empty_cache()
-
-        return epoch_loss / hyp_params.n_train
-
-    def evaluate(model, criterion, local_rank, test=False, train=False):
```

Comment:
move inner-function out of the main function. To reduce the size of main function.